README.md
# PDAJ
Python projects


pdaj/CHANGELOG
export_beam_integrals 1.0.2 (August 28, 2019)

    * Migrated from Bitbucket (Mercurial) to GitHub (Git).

export_beam_integrals 1.0.1 (April 11, 2016)

    * Vagrant: Added the vbguest plugin as a dependency, to keep VirtualBox
      Guest Additions up to date.

    * Vagrantfile: The configuration `docker.version' has been deprecated.

export_beam_integrals 1.0.0 (November 17, 2015)

    * Initial release.


pend.py
import sys
import numpy as np
from scipy.integrate import odeint

# The gravitational acceleration (m.s-2).
g = 9.81


def deriv(y, t, L1, L2, m1, m2):
    theta1, z1, theta2, z2 = y

    c, s = np.cos(theta1 - theta2), np.sin(theta1 - theta2)

    theta1dot = z1
    z1dot = (m2 * g * np.sin(theta2) * c - m2 * s * (L1 * z1 ** 2 * c + L2 * z2 ** 2) - (m1 + m2) * g * np.sin(
        theta1)) / L1 / (m1 + m2 * s ** 2)
    theta2dot = z2
    z2dot = ((m1 + m2) * (
                L1 * z1 ** 2 * s - g * np.sin(theta2) + g * np.sin(theta1) * c) + m2 * L2 * z2 ** 2 * s * c) / L2 / (
                        m1 + m2 * s ** 2)
    return theta1dot, z1dot, theta2dot, z2dot


def solve(L1, L2, m1, m2, tmax, dt, y0):
    t = np.arange(0, tmax + dt, dt)

    # Do the numerical integration of the equations of motion
    y = odeint(deriv, y0, t, args=(L1, L2, m1, m2))

    # Unpack z and theta as a function of time
    theta1, theta2 = y[:, 0], y[:, 2]

    # Convert to Cartesian coordinates of the two bob positions.
    x1 = L1 * np.sin(theta1)
    y1 = -L1 * np.cos(theta1)
    x2 = x1 + L2 * np.sin(theta2)
    y2 = y1 - L2 * np.cos(theta2)

    # print(theta1[-1], theta2[-1])
    return theta1[-1], theta2[-1], x1, y1, x2, y2


def gen_theta1_init_theta1_init(theta_resolution):
    search_space = np.linspace(0, 2 * np.pi, theta_resolution)
    for theta1_init in search_space:
        for theta2_init in search_space:
            yield theta1_init, theta2_init


def simulate_pendulum(theta_resolution):
    # Pendulum rod lengths (m), bob masses (kg).
    L1, L2 = 1, 1
    m1, m2 = 1, 1

    # Maximum time, time point spacings and the time grid (all in s).
    tmax, dt = 30, 0.001

    for theta1_init, theta2_init in gen_theta1_init_theta1_init(theta_resolution):
        y0 = np.array([theta1_init, 0, theta1_init, 0])
        theta1, theta2, x1, y1, x2, y2 = solve(L1, L2, m1, m2, tmax, dt, y0)
        yield theta1_init, theta2_init, theta1, theta2


def main():
    for theta1_init, theta2_init, theta1, theta2 in simulate_pendulum(15):
        print("%f, %f, %f, %f" % (theta1_init, theta2_init, theta1, theta2))


if __name__ == "__main__":
    main()

sequential.py
import argparse
import numpy as np
from scipy.integrate import odeint
import csv

DEFAULT_RESOLUTION = 6
DEFAULT_TMAX = 30
DEFAULT_DT = 0.01
DEFAULT_L1 = 1
DEFAULT_L2 = 1
DEFAULT_M1 = 1
DEFAULT_M2 = 1

# The gravitational acceleration (m.s-2).
g = 9.81


def deriv(y, t, L1, L2, m1, m2):
    """Return the first derivatives of y = theta1, z1, theta2, z2."""
    theta1, z1, theta2, z2 = y

    c, s = np.cos(theta1 - theta2), np.sin(theta1 - theta2)

    theta1dot = z1
    z1dot = (m2 * g * np.sin(theta2) * c - m2 * s * (L1 * z1 ** 2 * c + L2 * z2 ** 2) -
             (m1 + m2) * g * np.sin(theta1)) / L1 / (m1 + m2 * s ** 2)
    theta2dot = z2
    z2dot = ((m1 + m2) * (L1 * z1 ** 2 * s - g * np.sin(theta2) + g * np.sin(theta1) * c) +
             m2 * L2 * z2 ** 2 * s * c) / L2 / (m1 + m2 * s ** 2)
    return theta1dot, z1dot, theta2dot, z2dot


def solve(L1, L2, m1, m2, tmax, dt, y0):
    t = np.arange(0, tmax + dt, dt)
    # Do the numerical integration of the equations of motion
    y = odeint(deriv, y0, t, args=(L1, L2, m1, m2))
    theta1, theta2 = y[:, 0], y[:, 2]
    # Convert to Cartesian coordinates of the two bob positions.
    x1 = L1 * np.sin(theta1)
    y1 = -L1 * np.cos(theta1)
    x2 = x1 + L2 * np.sin(theta2)
    y2 = y1 - L2 * np.cos(theta2)

    return theta1, theta2, x1, y1, x2, y2


def gen_simulation_model_params(theta_resolution):
    search_space = np.linspace(0, 2 * np.pi, theta_resolution)
    for theta1_init in search_space:
        for theta2_init in search_space:
            yield theta1_init, theta2_init


def simulate_pendulum(theta_resolution, dt=DEFAULT_DT, tmax=DEFAULT_TMAX, L1=DEFAULT_L1, L2=DEFAULT_L2, m1=DEFAULT_M1,
                      m2=DEFAULT_M2):
    for theta1_init, theta2_init in gen_simulation_model_params(theta_resolution):
        # Initial conditions: theta1_init, dtheta1_init/dt, theta2_init, dtheta2_init/dt.
        y0 = np.array([theta1_init, 0, theta2_init, 0])
        theta1, theta2, x1, y1, x2, y2 = solve(L1, L2, m1, m2, tmax, dt, y0)
        yield theta1_init, theta2_init, theta1, theta2, x1, y1, x2, y2


def calculateAndWrite():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        'results_file',
        help="Filename where the results will be stored, in CSV format"
    )
    parser.add_argument(
        '-r',
        '--resolution',
        metavar='NUM',
        type=int,
        default=DEFAULT_RESOLUTION,
        help="Resolution, %d by default" % DEFAULT_RESOLUTION
    )
    parser.add_argument(
        '--tmax',
        metavar='NUM',
        type=float,
        default=DEFAULT_TMAX,
        help="Simulation time, %f by default" % DEFAULT_TMAX
    )
    parser.add_argument(
        '--dt',
        metavar='NUM',
        type=float,
        default=DEFAULT_DT,
        help="Simulation time step, %f by default" % DEFAULT_DT
    )
    args = parser.parse_args()
    results = simulate_pendulum(
        theta_resolution=args.resolution,
        dt=args.dt,
        tmax=args.tmax,
    )
    writeToCsv(results, args.results_file)


def writeToCsv(results, file):
    with open(file, 'w', newline='') as file:
        writer = csv.writer(file, delimiter=',', quotechar='|')
        writer = csv.DictWriter(file, fieldnames=["theta1_init", "theta2_init", "theta1", "theta2"])
        writer.writeheader()
        for r in results:
            theta1_init, theta2_init, theta1, theta2, x1, y1, x2, y2 = r
            writer.writerow({'theta1_init': theta1_init, 'theta2_init': theta2_init, 'theta1': theta1[len(theta1) - 1],
                             'theta2': theta2[len(theta2) - 1]})


if __name__ == '__main__':
    calculateAndWrite()


pdaj/pdaj/celeryconfig.py
from multiprocessing import cpu_count
import os

from beam_integrals import DEFAULT_MAX_MODE, DEFAULT_DECIMAL_PRECISION
from kombu import Queue

## Environment based settings

MAX_CPU_CORES = os.getenv('MAX_CPU_CORES', cpu_count())
SERVER_NAME = os.getenv('SERVER_NAME', 'localhost')
AM_I_SERVER = (os.getenv('COMPUTER_TYPE') == 'server')

PDAJ_MAX_MODE = int(os.getenv('PDAJ_MAX_MODE', DEFAULT_MAX_MODE))
PDAJ_DECIMAL_PRECISION = int(os.getenv('PDAJ_DECIMAL_PRECISION', DEFAULT_DECIMAL_PRECISION))
RESULTS_DIR = os.getenv('RESULTS_DIR', '/tmp/results')
STATUS_DIR = os.path.join(RESULTS_DIR, 'status')

## Concurrency settings

CELERYD_CONCURRENCY = MAX_CPU_CORES

# This ensures that each worker will only take one task at a time, when combined
# with late acks. This is the recommended configuration for long-running tasks.
# References:
#   * http://celery.readthedocs.org/en/latest/userguide/optimizing.html#prefetch-limits
#   * http://celery.readthedocs.org/en/latest/userguide/optimizing.html#reserve-one-task-at-a-time
#   * http://celery.readthedocs.org/en/latest/configuration.html#celeryd-prefetch-multiplier
#   * http://stackoverflow.com/questions/16040039/understanding-celery-task-prefetching
#   * https://bugs.launchpad.net/openquake-old/+bug/1092050
#   * https://wiredcraft.com/blog/3-gotchas-for-celery/
#   * http://www.lshift.net/blog/2015/04/30/making-celery-play-nice-with-rabbitmq-and-bigwig/
CELERYD_PREFETCH_MULTIPLIER = 1

## Task result backend settings

CELERY_RESULT_BACKEND = "redis://%s" % SERVER_NAME

## Message Routing

CELERY_DEFAULT_QUEUE = 'worker'
CELERY_DEFAULT_EXCHANGE = 'tasks'
CELERY_DEFAULT_ROUTING_KEY = 'worker'

PDAJ_RESOLUTION = os.getenv('RESOLUTION', 60)
PDAJ_L1 = os.getenv('L1', 1)
PDAJ_L2 = os.getenv('L2', 1)
PDAJ_TMAX = os.getenv('TMAX', 30)
PDAJ_DT = os.getenv('D1', 0.01)
PDAJ_M1 = os.getenv('M1', 1)
PDAJ_M2 = os.getenv('M2', 1)


if AM_I_SERVER:
    CELERY_QUEUES = (
        Queue('server', routing_key='server'),
    )
else:
    CELERY_QUEUES = (
        Queue('worker', routing_key='worker'),
    )


class ServerTasksRouter(object):
    def route_for_task(self, task, args=None, kwargs=None):
        if task.startswith('pdaj.tasks.server.'):
            return {'queue': 'server'}

        return None


CELERY_ROUTES = (
    ServerTasksRouter(),
)

## Broker Settings

BROKER_URL = "amqp://%s" % SERVER_NAME
CELERY_ACCEPT_CONTENT = ['pickle', 'json']

## Task execution settings

CELERY_MESSAGE_COMPRESSION = 'bzip2'
CELERY_TASK_RESULT_EXPIRES = None
CELERY_DISABLE_RATE_LIMITS = True
CELERY_TRACK_STARTED = True

# This ensures that the worker acks the task *after* it's completed.
# If the worker crashes or gets killed mid-execution, the task will be returned
# to the broker and restarted on another worker.
# References:
#   * https://wiredcraft.com/blog/3-gotchas-for-celery/
#   * http://celery.readthedocs.org/en/latest/configuration.html#celery-acks-late
#   * http://celery.readthedocs.org/en/latest/faq.html#faq-acks-late-vs-retry
CELERY_ACKS_LATE = True

## Worker settings

if AM_I_SERVER:
    CELERY_IMPORTS = ['pdaj.tasks.server']
else:
    CELERY_IMPORTS = ['pdaj.tasks.worker']

# HACK: Prevents weird SymPy related memory leaks
CELERYD_MAX_TASKS_PER_CHILD = 10


pdaj/pdaj/tasks/server.py
from datetime import datetime
import os
import csv
from celery import chord
from celery.exceptions import Reject
import numpy as np
from ..app import app
from .worker import solve


@app.task
def writeToCsv(results):
    file = app.conf.RESULTS_DIR + "/distributed.csv"
    with open(file, 'w') as file:
        writer = csv.writer(file, delimiter=',', quotechar='|')
        writer = csv.DictWriter(file, fieldnames=["theta1_init", "theta2_init", "theta1", "theta2"])
        writer.writeheader()
        for r in results:
            t1, t2, theta1, theta2, x1, y1, x2, y2 = r
            writer.writerow({'theta1_init': t1, 'theta2_init': t2, 'theta1': theta1[len(theta1) - 1],
                             'theta2': theta2[len(theta2) - 1]})


def get_experiment_status_filename(status):
    return os.path.join(app.conf.STATUS_DIR, status)


def get_experiment_status_time():
    """Get the current local date and time, in ISO 8601 format (microseconds and TZ removed)"""
    return datetime.now().replace(microsecond=0).isoformat()


@app.task
def record_experiment_status(status):
    with open(get_experiment_status_filename(status), 'w') as fp:
        fp.write(get_experiment_status_time() + '\n')


## Seeding the computations

def gen_simulation_model_params(theta_resolution):
    search_space = np.linspace(0, 2 * np.pi, theta_resolution)
    for theta1_init in search_space:
        for theta2_init in search_space:
            yield theta1_init, theta2_init, app.conf.PDAJ_L1, \
                  app.conf.PDAJ_L2, app.conf.PDAJ_M1, app.conf.PDAJ_M2, app.conf.PDAJ_TMAX, app.conf.PDAJ_DT, \
                  np.array([theta1_init, 0, theta2_init, 0])


@app.task
def seed_computations(ignore_result=True):
    if os.path.exists(get_experiment_status_filename('started')):
        raise Reject('Computations have already been seeded!')
    record_experiment_status.si('started').delay()
    chord(
        (solve.s(theta1_init, theta2_init, L1, L2, M1, M2, TMAX, DT, y0)
         for theta1_init, theta2_init, L1, L2, M1, M2, TMAX, DT, y0 in
         gen_simulation_model_params(app.conf.PDAJ_RESOLUTION)),
        (writeToCsv.s())
    ).delay()
    record_experiment_status.si('completed').delay()


pdaj/pdaj/tasks/worker.py
from ..app import app
import numpy as np
from scipy.integrate import odeint

SCALE_FACTOR_HELPER = 2.

g = 9.81


def deriv(y, t, L1, L2, m1, m2):
    """Return the first derivatives of y = theta1, z1, theta2, z2."""
    theta1, z1, theta2, z2 = y
    c, s = np.cos(theta1 - theta2), np.sin(theta1 - theta2)
    theta1dot = z1
    z1dot = (m2 * g * np.sin(theta2) * c - m2 * s * (L1 * z1 ** 2 * c + L2 * z2 ** 2) -
             (m1 + m2) * g * np.sin(theta1)) / L1 / (m1 + m2 * s ** 2)
    theta2dot = z2
    z2dot = ((m1 + m2) * (L1 * z1 ** 2 * s - g * np.sin(theta2) + g * np.sin(theta1) * c) +
             m2 * L2 * z2 ** 2 * s * c) / L2 / (m1 + m2 * s ** 2)
    return theta1dot, z1dot, theta2dot, z2dot


@app.task
def solve(theta1_init, theta2_init, L1, L2, m1, m2, tmax, dt, y0):
    t = np.arange(0, tmax + dt, dt)
    # Do the numerical integration of the equations of motion
    y = odeint(deriv, y0, t, args=(L1, L2, m1, m2))
    theta1, theta2 = y[:, 0], y[:, 2]
    # Convert to Cartesian coordinates of the two bob positions.
    x1 = L1 * np.sin(theta1)
    y1 = -L1 * np.cos(theta1)
    x2 = x1 + L2 * np.sin(theta2)
    y2 = y1 - L2 * np.cos(theta2)
    return theta1_init, theta2_init, theta1, theta2, x1, y1, x2, y2

pdaj/docker/docker-configure.py
#!/usr/bin/env python
import os
from beam_integrals import DEFAULT_MAX_MODE, DEFAULT_DECIMAL_PRECISION
from beam_integrals.beam_types import BaseBeamType
from beam_integrals.characteristic_equation_solvers import best_roots_cache
from beam_integrals.exceptions import BestRootsCacheError



def verify_best_roots_cache(max_mode, decimal_precision):
    try:
        print ("Checking best roots cache for max_mode=%d, decimal_precision=%d..." % (max_mode, decimal_precision))
        for beam_type in BaseBeamType.plugins.instances_sorted_by_id:
            best_roots_cache.get(beam_type, max_mode, decimal_precision)

        print ('Cache OK.')
    except BestRootsCacheError, e: # Regen if there's any kind of a problem with the cache
        # The original exception message is too verbose
        short_message = e.message.split(" You'll need to regenerate the cache", 1)[0]
        print ("Cache is invalid: %s" % short_message)

        best_roots_cache.regenerate(max_mode, decimal_precision)
        print ('Cache regenerated.')

def main():
    verify_best_roots_cache(
        max_mode=int(os.getenv('BEAM_INTEGRALS_MAX_MODE', DEFAULT_MAX_MODE)),
        decimal_precision=int(os.getenv('BEAM_INTEGRALS_DECIMAL_PRECISION', DEFAULT_DECIMAL_PRECISION))
    )

if __name__ == '__main__':
    main()


parallel.py
import argparse
from multiprocessing import Pool
import numpy as np
from scipy.integrate import odeint
import csv

DEFAULT_RESOLUTION = 6
DEFAULT_TMAX = 30
DEFAULT_DT = 0.01
DEFAULT_L1 = 1
DEFAULT_L2 = 1
DEFAULT_M1 = 1
DEFAULT_M2 = 1

# The gravitational acceleration (m.s-2).
g = 9.81


def deriv(y, t, L1, L2, m1, m2):
    """Return the first derivatives of y = theta1, z1, theta2, z2."""
    theta1, z1, theta2, z2 = y

    c, s = np.cos(theta1 - theta2), np.sin(theta1 - theta2)

    theta1dot = z1
    z1dot = (m2 * g * np.sin(theta2) * c - m2 * s * (L1 * z1 ** 2 * c + L2 * z2 ** 2) -
             (m1 + m2) * g * np.sin(theta1)) / L1 / (m1 + m2 * s ** 2)
    theta2dot = z2
    z2dot = ((m1 + m2) * (L1 * z1 ** 2 * s - g * np.sin(theta2) + g * np.sin(theta1) * c) +
             m2 * L2 * z2 ** 2 * s * c) / L2 / (m1 + m2 * s ** 2)
    return theta1dot, z1dot, theta2dot, z2dot


def solve(L1, L2, m1, m2, tmax, dt, y0, t1init, t2init):
    t = np.arange(0, tmax + dt, dt)
    # Do the numerical integration of the equations of motion
    y = odeint(deriv, y0, t, args=(L1, L2, m1, m2))
    theta1, theta2 = y[:, 0], y[:, 2]
    # Convert to Cartesian coordinates of the two bob positions.
    x1 = L1 * np.sin(theta1)
    y1 = -L1 * np.cos(theta1)
    x2 = x1 + L2 * np.sin(theta2)
    y2 = y1 - L2 * np.cos(theta2)

    return theta1, theta2, x1, y1, x2, y2, t1init, t2init


def gen_simulation_model_params(theta_resolution):
    search_space = np.linspace(0, 2 * np.pi, theta_resolution)
    for theta1_init in search_space:
        for theta2_init in search_space:
            yield theta1_init, theta2_init, np.array([theta1_init, 0, theta2_init, 0])


def myFunc(args):
    return solve(DEFAULT_L1, DEFAULT_L2, DEFAULT_M1, DEFAULT_M2, DEFAULT_TMAX, DEFAULT_DT, args[2], args[0], args[1])


def simulate_pendulum(theta_resolution, dt=DEFAULT_DT, tmax=DEFAULT_TMAX, L1=DEFAULT_L1, L2=DEFAULT_L2, m1=DEFAULT_M1, m2=DEFAULT_M2):
    with Pool() as pool:
        # kao parametar prosledjena je funkcija koja predstavlja generator pocetnih uglova i y0
        results = pool.imap(myFunc, gen_simulation_model_params(theta_resolution), chunksize=theta_resolution)
        for res in results:
            # yildujem theta1,theta2,theta1_init i theta2_init
            yield res[0][-1], res[1][-1], res[6], res[7]


def calculateAndWrite():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        'results_file',
        help="Filename where the results will be stored, in CSV format"
    )
    parser.add_argument(
        '-r',
        '--resolution',
        metavar='NUM',
        type=int,
        default=DEFAULT_RESOLUTION,
        help="Resolution, %d by default" % DEFAULT_RESOLUTION
    )
    parser.add_argument(
        '--tmax',
        metavar='NUM',
        type=float,
        default=DEFAULT_TMAX,
        help="Simulation time, %f by default" % DEFAULT_TMAX
    )
    parser.add_argument(
        '--dt',
        metavar='NUM',
        type=float,
        default=DEFAULT_DT,
        help="Simulation time step, %f by default" % DEFAULT_DT
    )
    args = parser.parse_args()
    results = simulate_pendulum(
        theta_resolution=args.resolution,
        dt=args.dt,
        tmax=args.tmax,
    )
    writeToCsv(results, args.results_file)


def writeToCsv(results, file):
    with open(file, 'w', newline='') as file:
        writer = csv.writer(file, delimiter=',', quotechar='|')
        writer = csv.DictWriter(file, fieldnames=["theta1_init", "theta2_init", "theta1", "theta2"])
        writer.writeheader()
        for r in results:
            theta1, theta2, theta1_init, theta2_init = r
            writer.writerow({'theta1_init': theta1_init, 'theta2_init': theta2_init, 'theta1': theta1,
                             'theta2': theta2})
if __name__ == '__main__':
    calculateAndWrite()
